core/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -346,55 +343,44 @@ public Repository repository(String repository) {
+            RepositoryMetaData previousMetadata = previous.getMetadata();
+            if (!previousMetadata.type().equals(repositoryMetaData.type()) && previousMetadata.settings().equals(repositoryMetaData.settings())) {
                 // Previous version is the same as this one - ignore it
                 return false;
             }
         }
-        RepositoryHolder holder = createRepositoryHolder(repositoryMetaData);
+        Repository newRepo = createRepository(repositoryMetaData);
         if (previous != null) {
-            // Closing previous version
-            closeRepository(repositoryMetaData.name(), previous);
+            closeRepository(previous);
         }
-        Map<String, RepositoryHolder> newRepositories = new HashMap<>(repositories);
-        newRepositories.put(repositoryMetaData.name(), holder);
+        Map<String, Repository> newRepositories = new HashMap<>(repositories);
+        newRepositories.put(repositoryMetaData.name(), newRepo);
+        repositories = newRepositories;
         return true;
     }
 
-    /**
-     * Closes the repository
-     *
-     * @param name   repository name
-     * @param holder repository holder
-     */
-    private void closeRepository(String name, RepositoryHolder holder) throws IOException {
-        logger.debug("closing repository [{}][{}]", holder.type, name);
-        if (holder.repository != null) {
-            holder.repository.close();
-        }
+    /** Closes the given repository. */
+    private void closeRepository(Repository repository) throws IOException {
+        logger.debug("closing repository [{}][{}]", repository.getMetadata().type(), repository.getMetadata().name());
+        repository.close();
     }
 
     /**
      * Creates repository holder
      */
-    private RepositoryHolder createRepositoryHolder(RepositoryMetaData repositoryMetaData) {
+    private Repository createRepository(RepositoryMetaData repositoryMetaData) {
         logger.debug("creating repository [{}][{}]", repositoryMetaData.type(), repositoryMetaData.name());
-        Injector repositoryInjector = null;
+        Repository.Factory factory = typesRegistry.get(repositoryMetaData.type());
+        if (factory == null) {
+            throw new RepositoryException(repositoryMetaData.name(),
+                "repository type [" + repositoryMetaData.type() + "] does not exist");
+        }
         try {
-            ModulesBuilder modules = new ModulesBuilder();
-            RepositoryName name = new RepositoryName(repositoryMetaData.type(), repositoryMetaData.name());
-            modules.add(b -> {
-                b.bind(RepositoryName.class).toInstance(name);
-                typesRegistry.bindType(b, repositoryMetaData.type());
-                b.bind(RepositorySettings.class).toInstance(new RepositorySettings(settings, repositoryMetaData.settings()));
-            });
-
-            repositoryInjector = modules.createChildInjector(injector);
```

Comment:
since we own that code :( can we remove that createChildInjector method as well or throw UOE?